main/mpu9250.go

```diff
@@ -0,0 +1,182 @@
+
+	time.Sleep(100 * time.Microsecond)
+
+	r, err := i2cbus.ReadByteFromReg(0x68, 0x49)
+	chkErr(err)
+
+	ret := r == 0x48
+
+	// Read calibration data.
+	setSetting(0x25, 0x0C|0x80)
+	setSetting(0x26, 0x10)
+	setSetting(0x27, 0x83) // Read three bytes, (CalX, CalY, CalZ).
+
+	mxcal, err := i2cbus.ReadByteFromReg(0x68, 0x49)
+	chkErr(err)
+	mycal, err := i2cbus.ReadByteFromReg(0x68, 0x4A)
+	chkErr(err)
+	mzcal, err := i2cbus.ReadByteFromReg(0x68, 0x4B)
+	chkErr(err)
+
+	magXcal = (float64(mxcal)-128)/256.0 + 1.0
+	magYcal = (float64(mycal)-128)/256.0 + 1.0
+	magZcal = (float64(mzcal)-128)/256.0 + 1.0
+
+	return ret
+}
+
+func initMPU9250() {
+	initI2C()
+	globalSettings.AHRS_Enabled = true
+	mySituation.mu_Attitude = &sync.Mutex{}
+
+	//TODO: Calibration.
+
+	setSetting(0x6B, 0x80) // Reset.
+	time.Sleep(100 * time.Millisecond)
+	setSetting(0x6B, 0x01) // Clock source.
+	setSetting(0x6C, 0x00) // Enable accelerometer and gyro.
+
+	setSetting(0x6A, 0x20) // I2C Master mode.
+	setSetting(0x24, 0x0D) // I2C configuration multi-master, 400KHz.
+
+	// AK8963 init.
+
+	setSetting(0x25, 0x0C) // Set the I2C slave addres of AK8963 and set for write.
+	setSetting(0x26, 0x0B) // I2C slave 0 register address from where to begin data transfer.
+	setSetting(0x63, 0x01) // Reset AK8963.
+	setSetting(0x27, 0x81) // Enable I2C and set 1 byte.
+
+	setSetting(0x25, 0x0C) // Set the I2C slave addres of AK8963 and set for write.
+	setSetting(0x26, 0x0A) // I2C slave 0 register address from where to begin data transfer.
+	setSetting(0x63, 0x16) // Register value to 100Hz continuous measurement in 16bit.
+	setSetting(0x27, 0x81) // Enable I2C and set 1 byte.
+
+	// Accelerometer and gyro init.
+
+	setSetting(0x19, 0x00) // Set Gyro 1000 Hz sample rate. rate = gyroscope output rate/(1 + value)
+	setSetting(0x1A, 0x03) // Set low pass filter to 92 Hz.
+	setSetting(0x1B, 0x00) // Set gyro sensitivity to 250dps.
+	setSetting(0x1C, 0x00) // Set accelerometer scale to +/- 2G.
+	setSetting(0x1D, 0x02) // Set Accel 1000 Hz sample rate.
+
+	if !checkMagConnection() {
+		log.Printf("magnetometer is offline.\n")
+		return
+	}
+
+	go readRawData()
+	go calculateAttitude()
+}
+
+func readRawData() {
+	timer := time.NewTicker(2 * time.Millisecond)
+
+	for {
+		<-timer.C
+		// Get accelerometer data.
+		x_acc, err := i2cbus.ReadWordFromReg(0x68, 0x3B)
+		chkErr(err)
+		y_acc, err := i2cbus.ReadWordFromReg(0x68, 0x3D)
+		chkErr(err)
+		z_acc, err := i2cbus.ReadWordFromReg(0x68, 0x3F)
+
+		// currently manually setting resolution
+		x_acc_f := float64(int16(x_acc)) * 0.00006103515625
+		y_acc_f := float64(int16(y_acc)) * 0.00006103515625
+		z_acc_f := float64(int16(z_acc)) * 0.00006103515625
+
+		// Get gyro data.
+		x_gyro, err := i2cbus.ReadWordFromReg(0x68, 0x43)
+		chkErr(err)
+		y_gyro, err := i2cbus.ReadWordFromReg(0x68, 0x45)
+		chkErr(err)
+		z_gyro, err := i2cbus.ReadWordFromReg(0x68, 0x47)
+
+		x_gyro_f := float64(int16(x_gyro)) * 0.00006103515625
```

Comment:
Gyro scale factor of 0.000061035 deg per LSB is incorrect for ±250 deg/sec full scale -- looks like a copy / paste from the accelerometer scaling. This will understate gyro motion by a factor of 125.

Correct scale factor is 1/131 deg per LSB per the MPU9250 spec sheet. However, even this will be off by a factor of Pi when the Madgwick method integrates the angle rates -- e.g. if you rotate the sensor by 45°, it will only show a 14.3° displacement. Suggest using Pi/131 as a scale factor until the source of this discrepancy can be found:

```
x_gyro_f := float64(int16(x_gyro)) * math.Pi / 131.0
y_gyro_f := float64(int16(y_gyro)) * math.Pi / 131.0
z_gyro_f := float64(int16(z_gyro)) * math.Pi / 131.0
```